### PR TITLE
Jg 541 subset spatial update

### LINKED ISSUE #507: subset_spatial op should work with valid polygons
### Expected behavior

`subset_spatial` should work with all valid polygons regardless of their actual spatial extent. 

### Actual behavior

If I invoke `subset_spatial` with `ds`:

    esacci.2010.CLOUD.mon.L3C.CLD_PRODUCTS.multi-sensor.multi-platform.ATSR2-AATSR.2-0.r1

and `region` WKT defining the following, rather small polygon from ESA CCI Glaciers dataset:

    "POLYGON ((-53.57632337894747 67.10673414497307, -53.57626614693088 67.10619646968784, -53.575575794544775 67.10620761160448, -53.57551858059012 67.10566993596773, -53.57482824280637 67.1056810746106, -53.57479964451354 67.10541223662418, -53.574109313685 67.10542337213865, -53.57408072373028 67.1051545339918, -53.572009750037914 67.10518792220135, -53.571981183714094 67.10491908360363, -53.56991022674504 67.10495244453077, -53.569881684051325 67.10468360548275, -53.56919137132743 67.10469471968085, -53.56921990637478 67.10496355887388, -53.56852958531452 67.1049746702339, -53.56855811340513 67.10524350955681, -53.5671774539219 67.10526572361778, -53.56712042901686 67.10472804437723, -53.56781074381009 67.10471693912801, -53.56778222474533 67.10444809962996, -53.568472531202524 67.10443699154277, -53.56838695520772 67.10363047339246, -53.569077238037295 67.10361936275717, -53.56904870643948 67.10335052348844, -53.56973898093374 67.10333941001542, -53.569710442380476 67.10307057087647, -53.57040070853927 67.10305945456565, -53.570372163030925 67.1027906155566, -53.571062420854446 67.1027794964081, -53.57103386839113 67.10251065752905, -53.57241436667729 67.10248841057422, -53.57238579961488 67.10221957197018, -53.57445651892933 67.10218617960403, -53.57442792962403 67.10191734142042, -53.575808390422566 67.1018950652207, -53.575779786519966 67.10162622731258, -53.57785044959804 67.10159279108024, -53.57782182345493 67.10132395359305, -53.57920224675896 67.1013016481501, -53.57917360602069 67.10103281093878, -53.58262461394139 67.10097699586724, -53.58259593567808 67.10070815936882, -53.58604688804876 67.10065227046755, -53.58607560452903 67.10092110623687, -53.5867658005646 67.10090991936381, -53.58679452538237 67.10117875497203, -53.587484728366974 67.10116756497067, -53.5874559959055 67.10089872950844, -53.588146190552095 67.10088753667073, -53.58805997440654 67.10008103063063, -53.58875014542906 67.10006983524875, -53.58877889109226 67.10033867046462, -53.589469069062815 67.10032747195446, -53.589555333151964 67.10113397707244, -53.58817493065678 67.10115637198695, -53.588203671456405 67.10142520728797, -53.58751346152288 67.10143640041765, -53.58754219537351 67.10170523584945, -53.58685197710164 67.10171642614279, -53.586880704002844 67.10198526170532, -53.58619047739239 67.10199644916216, -53.58621919734403 67.10226528485546, -53.58414849041439 67.10229882976869, -53.58417718812709 67.1025676658843, -53.58348694345156 67.10257884170274, -53.583515634213306 67.10284767794896, -53.582825381199484 67.1028588509306, -53.58288274951425 67.10339652366906, -53.58357301781828 67.10338535039598, -53.583601710661156 67.10365418659654, -53.585672534345264 67.10362064844344, -53.58570125081805 67.10388948419128, -53.58639153163644 67.10387829869555, -53.586420256449124 67.10414713428231, -53.585729967985365 67.10415831992393, -53.58575868584685 67.10442715564143, -53.58506838904291 67.10443833844619, -53.58509709995305 67.10470717429435, -53.58371648897093 67.10472953124703, -53.58368779335269 67.10446069510724, -53.582307194887086 67.10448303983686, -53.58233587521329 67.1047518762682, -53.58164556729428 67.10476304430442, -53.58167424066754 67.10503188086624, -53.580983924408926 67.10504304606525, -53.58101259082931 67.10531188275765, -53.58032226623112 67.1053230451192, -53.580350925698 67.10559188194198, -53.579660592760035 67.10560304146625, -53.57971789847896 67.10614071535736, -53.579027549554304 67.1061518721897, -53.57908484274993 67.10668954631129, -53.57839447783797 67.10670070045153, -53.57842311782737 67.10696953763514, -53.57704237063045 67.10699183725701, -53.57701374593649 67.10672299978249, -53.57632337894747 67.10673414497307, -53.57632337894747 67.10673414497307))"

I get `IndexError: index 1 is out of bounds for axis 0 with size 0`:

    Cate Desktop, version 1.0.1-dev.1

    set_workspace_resource() call raised exception: "index 1 is out of bounds for axis 0 with size 0"

    An error (code 20) occurred in Cate Core:

    Traceback (most recent call last):
      File "D:\Projects\cate\cate\util\web\jsonrpchandler.py", line 192, in send_service_method_result
        result = future.result()
      File "D:\Miniconda3\envs\cate-env\lib\concurrent\futures\_base.py", line 398, in result
        return self.__get_result()
      File "D:\Miniconda3\envs\cate-env\lib\concurrent\futures\_base.py", line 357, in __get_result
        raise self._exception
      File "D:\Miniconda3\envs\cate-env\lib\concurrent\futures\thread.py", line 55, in run
        result = self.fn(*self.args, **self.kwargs)
      File "D:\Projects\cate\cate\util\web\jsonrpchandler.py", line 269, in call_service_method
        result = method(*method_params, monitor=monitor)
      File "D:\Projects\cate\cate\webapi\websocket.py", line 284, in set_workspace_resource
        monitor=monitor)
      File "D:\Projects\cate\cate\core\wsmanag.py", line 323, in set_workspace_resource
        workspace.execute_workflow(res_name=res_name, monitor=monitor)
      File "D:\Projects\cate\cate\core\workspace.py", line 600, in execute_workflow
        self.workflow.invoke_steps(steps, context=self._new_context(), monitor=monitor)
      File "D:\Projects\cate\cate\core\workflow.py", line 632, in invoke_steps
        step.invoke(context=context, monitor=monitor.child(work=1))
      File "D:\Projects\cate\cate\core\workflow.py", line 318, in invoke
        self._invoke_impl(_new_context(context, step=self), monitor=monitor)
      File "D:\Projects\cate\cate\core\workflow.py", line 980, in _invoke_impl
        return_value = self._op(monitor=monitor, **input_values)
      File "D:\Projects\cate\cate\core\op.py", line 215, in __call__
        return_value = self._wrapped_op(**input_values)
      File "D:\Projects\cate\cate\ops\subset.py", line 57, in subset_spatial
        return adjust_spatial_attrs(subset_spatial_impl(ds, region, mask))
      File "D:\Projects\cate\cate\core\op.py", line 215, in __call__
        return_value = self._wrapped_op(**input_values)
      File "D:\Projects\cate\cate\ops\normalize.py", line 78, in adjust_spatial_attrs
        return adjust_spatial_attrs_impl(ds)
      File "D:\Projects\cate\cate\util\opimpl.py", line 176, in adjust_spatial_attrs_impl
        geoattrs = _get_spatial_props(ds, dim)
      File "D:\Projects\cate\cate\util\opimpl.py", line 285, in _get_spatial_props
        dim_res = abs(ds[dim].values[1] - ds[dim].values[0])
    IndexError: index 1 is out of bounds for axis 0 with size 0

### Specifications 

Cate 1.0.1.dev1

### LINKED ISSUE #508: subset_spatial op very slow with non-trivial polygons
### Problem

`subset_spatial` op is *very* slow (too slow) with `mask=True` and `region` being a non-trivial polygons and when many grid cells have to be compared. 

This is probably because we compare for each grid cell individually if it intersects the polygon. Would be more efficient if we create the mask by rendering the polygon as a masked numpy array (binary image).

### Steps

In Cate Desktop:

1.  Select operation `read_necdf` and add step with with file name e.g. `"20100401120000-ESACCI-L4_GHRSST-SSTdepth-OSTIA-GLOB_LT-v02.0-fv01.1.nc"` 
2.  From countries layer select Germany (yes, SST is an ocean product, but still there are many coastal cells this way)
3.  Select operation `subset_spatial` and add step with `ds_1`, the selected POLYGON and mask on. 

This actually computes forever.

### Specifications 

Cate 1.1 dev1

### LINKED ISSUE #541: subset_spatial won't select 'long' rectangles
### Expected behavior

When trying to view the arctic area, I would like to select a long polygon on 'top' of the map and plot it in polar_stereographic projection.

### Actual behavior

When we convert PolygonLike to shapely polygons we loose the information regarding which value was ``lon_min``, which was ``lon_max``, as the smallest absolute value will be regarded as ``min``. E.g., if a user provides ``{'lon_min':170, 'lon_max':-170}``, intending to select 20 degrees, this will be converted to ``{'lon_min':-170, 'lon_max':170}``, selecting 160 degrees.

To work around this, a heuristic is used in subset_spatial, assuming that users typically don't select very large rectangles on the map, and usually the smallest version of two possibilities is the correct one.

When trying to select the polar region, this assumption fails.

### Steps to reproduce the problem

Try to use ``subset_spatial()'' with ``{'lon_min': -185, 'lon_max': 185}``, observe a small slice crossing the international date line being selected.

### Specifications 

Latest Cate dev

### LINKED ISSUE #559: plot_map data plot type
Check the picture:
cate_plots

Both plots are on the same grid. The first plot is current implementation, the second plot could be updated implementation. Should we change to the second type of plots, e.g., plot a colormesh on top of a map instead of a contour?

### LINKED ISSUE #560: Select all pixels crossed by a given polygon
### Expected behavior

As mentioned by @forman in #507, the operation should select all pixels crossed by the polygon.

### Actual behavior

Currently ``subset_spatial()`` only selects pixels whose centers fall inside the polygon:
subset_spatial_now
subset_spatial_now1

### PR DESCRIPTION
**Pull request by JanisGailis**
Quite important changes to ``subset_spatial_impl``. Also some rather minor changes in ``ds`` implementations to make sure we use the same implementation methods for the same tasks.

Closes #541 
We can now select as well as plot 'long' polygons. Using explicit coordinates instead of polygons should now 'always' work as expected.

Closes #508 
Performance of ``subset_spatial()`` is dramatically improved. Non-masked polygons will always use xarray selection machinery, Masked polygons first subset the dataset to polygon extents and then performs fast masking.

Closes #560 
``subset_spatial()`` now selects all pixels crossed by a given polygon. Both when xarray select is used, as well as with masking enabled.

Closes #507 
``subset_spatial()`` Now works with polygons smaller than pixels. Selecting either a single pixel, or in case it includes a pixel vertice or edge - crossed pixels. It also works when the provided dataset contains only a single pixel. If ``subset_spatial()`` is invoked with a polygon that does not cross the spatial extents of the given dataset, it will raise an appropriate error message.

Closes #559 
``plot_map()`` now by default produces a 'pixel' instead of a 'contour' plot. It is still possible to plot contours, by setting the appropriate parameter to 'True'.

Maybe something else I have forgotten.

### CODE REVIEW

**Comment by forman** [comment]
Thanks Janis. What a PR. Could you next time please limit PRs to one per issue, otherwise review takes too much time.

File: cate/core/opimpl.py [hunk 1]
```diff
@@ -407,45 +488,51 @@ def subset_spatial_impl(ds: xr.Dataset,
...
-    lon_min, lat_min, lon_max, lat_max = region.bounds
+    lon_min, lat_min, lon_max, lat_max = extents
 
     # Validate the bounding box
     if (not (-90 <= lat_min <= 90)) or \
             (not (-90 <= lat_max <= 90)) or \
             (not (-180 <= lon_min <= 180)) or \
             (not (-180 <= lon_max <= 180)):
-        raise ValueError('Provided polygon extent outside of geospatial'
+        raise ValueError('Provided polygon extends outside of geospatial'
```
Comment (forman) [inline cate/core/opimpl.py]: Please use "extent", see http://www.dictionary.com/browse/extent

Comment (JanisGailis) [reply cate/core/opimpl.py]: Ah... I don't agree. :)
The depends of you want a statement, or a description. But even if it's a statement, just 'extent' would be wrong-ish. So it's either:
'Given polygon extent is outside of bounds'

Or with a verb (http://www.dictionary.com/browse/extend)
'Provided polygon extends outside of bounds'

**Comment by JanisGailis** [comment]
Forgot to update CHANGES.md... Doing it now.

**Comment by JanisGailis** [comment]
@kbernat It'd be cool if you can take a look at the changes this brings to ``ds`` implementations. In a sentence - ``subset_spatial_impl`` now works on a ``PolygonLike`` and tries to be smart about it, so, there's no need to convert the ``region`` in ``ds`` classes, rather, the ``PolygonLike`` just gets passed on ``subset_spatial_impl`` instead. Also, use ``adjust_spatial_attrs`` instead of determining dataset extents manually, to keep the same concern in the same place.

If you're too busy today, I'll merge it anyway at the end of day and help you fix things in case this messes anything up (it shouldn't).

**Pull request merged**